commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/AbstractChunkIndex.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2023 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.tiered.storage.commons.chunkindex;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public abstract class AbstractChunkIndex implements ChunkIndex {
+    protected final int originalChunkSize;
+    protected final int originalFileSize;
+
+    protected int chunksAdded = 0;
+    protected boolean finished = false;
+    protected int finalTransformedChunkSize = -1;
+
+    protected AbstractChunkIndex(final int originalChunkSize, final int originalFileSize) {
+        checkSize(originalChunkSize, "Original chunk size");
+        this.originalChunkSize = originalChunkSize;
+
+        checkSize(originalFileSize, "Original file size");
+        this.originalFileSize = originalFileSize;
+    }
+
+    @Override
+    public final void addChunk(final int transformedChunkSize) {
+        if (finished) {
+            throw new IllegalStateException("Cannot add chunk to already finished index");
+        }
+
+        checkSize(transformedChunkSize, "Transformed chunk size");
+
+        // Check that we expect only the final chunk at this point.
+        if (remainOfOriginalFileSize() <= originalChunkSize) {
+            throw new IllegalStateException("This must be final chunk. Call `finish` instead.");
+        }
+
+        addChunk0(transformedChunkSize);
+
+        this.chunksAdded += 1;
+    }
+
+    protected void addChunk0(final int transformedChunkSize) {}
+
+    @Override
+    public final void finish(final int finalTransformedChunkSize) {
+        if (finished) {
+            throw new IllegalStateException("Cannot finish already finished index");
+        }
```

Comment:
This whole mechanics around finishing an index asks for a separate builder class, but probably a bit later